Speaker Feedback: Add check for whether a session is currently accepting feedback

coreymckrill opened the pull request:
This adds a function to comprehensively check whether a post can and is currently accepting feedback comments. The checks include whether it's the right type of session, and close the current time is to the session's start time, plus some other things. This is useful in both the REST API and front end templates for determining what kind of response or content to return.

Both the REST API controller and the feedback form renderer are also updated to use this function.

Fixes #343, Fixes #394

### Screenshots

![free-lunch](https://user-images.githubusercontent.com/916023/78081775-38801d00-7366-11ea-9320-ba9ed4c1ab7d.jpg)

### How to test the changes in this Pull Request:

1. Try viewing the feedback form for a session post (i.e. `site.wordcamp.test/session-slug/feedback`)
1. If the session type is set to regular and the session time is slightly before the current time, you should see the feedback form. Otherwise, you should see a message explaining why the feedback form isn't available. Try changing those parameters to get different messages.
1. Try submitting feedback for that same session post using the [REST API](https://github.com/WordPress/wordcamp.org/pull/353). The same error messages should be in the API responses when the session's conditions aren't met.
1. All phpunit tests should pass.

coreymckrill committed:
Add check for accepting feedback, template for form errors

Adds a function to comprehensively check whether a post can and is
currently accepting feedback comments. This is useful in both the REST
API and front end templates for determining what kind of response or
content to return.

Fixes #343 Fixes #394

coreymckrill committed:
Fix copy pasta

coreymckrill committed:
Fix rest controller tests

coreymckrill committed:
Add tests for post_accepts_feedback

coreymckrill committed:
Use constant for form time window

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/form.php:
```diff
@@ -17,16 +27,52 @@ function has_feedback_form() {
+		$content = $content . ob_get_clean(); // Append form to the normal content.
 	} elseif ( is_page( get_option( OPTION_KEY ) ) ) {
+		$wordcamp   = get_wordcamp_post();
+		$start_date = $wordcamp->meta['Start Date (YYYY-mm-dd)'][0] ?? $now->getTimestamp() + YEAR_IN_SECONDS;
```
Why default this to a year from now, so it's an arbitrary future date, or is the year important?

ryelle replied:
I think the fact that the start date is not the real UTC value will bite some of the UTC+ timezones — they might have morning sessions, but `/feedback` will say the event hasn't started. We could convert this value into the site's timezone, since that has a good chance of being correct. Something like

```php
$date_string = date( 'Y-m-d', $wordcamp->meta['Start Date (YYYY-mm-dd)'][0] );
$start_date  = date_create( $date_string, wp_timezone() )->getTimestamp();
```

coreymckrill replied:
> Why default this to a year from now, so it's an arbitrary future date, or is the year important?

The thinking was that if the site doesn't have a meta value set for the start date, it probably hasn't been scheduled yet, in which case, the correct message would be `Feedback forms are not available until the event has started`. If the value defaulted to `0`, the message shown would be `Feedback forms are closed for this event`, which doesn't seem as accurate. The fact that it's a year ahead is arbitrary, it just needs to be far enough ahead that it won't fall into any window of time that we may set in the future.

coreymckrill replied:
> I think the fact that the start date is not the real UTC value will bite some of the UTC+ timezones

Good call. I expanded on your suggested here: 223b0f61

ryelle replied:
Looks good, and you were able to add comments which addressed my initial  🤔❓

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/post.php:
```diff
@@ -0,0 +1,66 @@
+<?php
```
Why add this as a new file/namespace, instead of adding to `form.php`? It feels like an extra abstraction to me.  (no change requested, just curious)

coreymckrill replied:
Two reasons:

1. The `post_accepts_feedback` function is also used in the REST API controller. Since it has multiple contexts, it seemed better to keep it separated from either context. That helps avoid the code making assumptions that are too specific to one particular context.
1. It seems possible that we will eventually have other functions that deal specifically with the posts that feedback comments are attached to. When I wrote this PR, I had at least one other in mind, but now I'm not thinking of it 😕

ryelle replied:
Makes sense! thanks for answering my curiosity 🙂

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/post.php:
```diff
@@ -0,0 +1,66 @@
+	if ( 'publish' !== get_post_status( $post ) ) {
+		return new WP_Error(
+			'speaker_feedback_post_unavailable',
+			__( 'This post does is not available for feedback.', 'wordcamporg' )
```
```suggestion
			__( 'This post is not available for feedback.', 'wordcamporg' )
```

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/form.php:
```diff
@@ -17,16 +27,52 @@ function has_feedback_form() {
 function render( $content ) {
+	global $post;
+
+	$now = date_create( time(), wp_timezone() );
```
```suggestion
	$now = date_create( 'now', wp_timezone() );
```

This was `false` on my local env, the docs for [DateTime](https://www.php.net/manual/en/datetime.construct.php) say you should use `"now"` to get the current time with a timezone.

coreymckrill committed:
Account for timezone when evaluating start date

For places with UTC+ timezones, using the default UTC start date could
result in the Feedback chooser page saying the event hasn't started
yet when it actually has.

coreymckrill committed:
Update public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/post.php

Co-Authored-By: Kelly Dwan <ryelle@users.noreply.github.com>

coreymckrill committed:
Update public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/form.php

Co-Authored-By: Kelly Dwan <ryelle@users.noreply.github.com>

ryelle reviewed (approved):
Looks good now 👍
